### PR TITLE
[BUGFIX] Fix `math.get_dtype_name` for `ArrayBox`

### LINKED ISSUE #4492: [BUG] Finite diff changes incompatible with floats
### Expected behavior

The following quantum function should be able to be run:

```python
import pennylane as qml

def f(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliY(0))

def interpretted_grad_default(x):
    device = qml.device("default.qubit", wires=1)
    g = qml.QNode(f, device, diff_method="finite-diff")
    h = qml.grad(g, argnum=0)
    return h(x)

print(interpretted_grad_default(1.0))
```

### Actual behavior

There's an assertion: `ImportError: autoray couldn't find function 'get_dtype_name' for backend 'autograd'.`

### Additional information

Catalyst tests are failing due to [a change made to finite diff](https://github.com/PennyLaneAI/pennylane/pull/4415#discussion_r1290048575). This needs to be addressed before the v0.32 release.

### Source code

See above

### Tracebacks

```
Traceback (most recent call last):
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 953, in get_dtype_name
    return x.dtype.name
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autograd/numpy/numpy_boxes.py", line 19, in <lambda>
    dtype = property(lambda self: self._value.dtype)
AttributeError: 'float' object has no attribute 'dtype'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 462, in get_lib_fn
    lib_fn = _FUNCS[backend, fn]
KeyError: ('autograd', 'get_dtype_name')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 429, in import_lib_fn
    lib_fn = _FUNCS[backend, fn] = wrapper(getattr(lib, fn_name))
AttributeError: module 'pennylane.numpy' has no attribute 'get_dtype_name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/erick.ochoalopez/Code/catalyst-latest/test.py", line 29, in <module>
    print(interpretted_grad_default(1.0))
  File "/home/erick.ochoalopez/Code/catalyst-latest/test.py", line 27, in interpretted_grad_default
    return h(x)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/_grad.py", line 120, in __call__
    grad_value, ans = grad_fn(*args, **kwargs)  # pylint: disable=not-callable
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autograd/wrap_util.py", line 20, in nary_f
    return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/_grad.py", line 146, in _grad_with_forward
    grad_value = vjp(vspace(ans).ones())
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autograd/core.py", line 14, in vjp
    def vjp(g): return backward_pass(g, end_node)
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autograd/core.py", line 21, in backward_pass
    ingrads = node.vjp(outgrad[0])
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autograd/core.py", line 67, in <lambda>
    return lambda g: (vjp(g),)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/interfaces/autograd.py", line 497, in grad_fn
    vjp_tapes, processing_fn = qml.gradients.batch_vjp(
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/gradients/vjp.py", line 521, in batch_vjp
    g_tapes, fn = vjp(tape, dy, gradient_fn, gradient_kwargs=gradient_kwargs)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/gradients/vjp.py", line 381, in vjp
    gradient_tapes, fn = gradient_fn(tape, **gradient_kwargs)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/transforms/batch_transform.py", line 359, in __call__
    return self._tape_wrapper(*targs, **tkwargs)(qnode)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/transforms/batch_transform.py", line 452, in <lambda>
    return lambda tape: self.construct(tape, *targs, **tkwargs)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/transforms/batch_transform.py", line 434, in construct
    tapes, processing_fn = self.transform_fn(tape, *targs, **tkwargs)
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/gradients/finite_difference.py", line 334, in finite_diff
    if any(qml.math.get_dtype_name(p) == "float32" for p in tape.get_parameters()):
  File "/home/erick.ochoalopez/Code/pennylane/pennylane/gradients/finite_difference.py", line 334, in <genexpr>
    if any(qml.math.get_dtype_name(p) == "float32" for p in tape.get_parameters()):
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 956, in get_dtype_name
    return do("get_dtype_name", x, like=x)
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 79, in do
    return get_lib_fn(backend, fn)(*args, **kwargs)
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 464, in get_lib_fn
    lib_fn = import_lib_fn(backend, fn)
  File "/home/erick.ochoalopez/Code/catalyst-latest/env/lib/python3.10/site-packages/autoray/autoray.py", line 437, in import_lib_fn
    raise ImportError(
ImportError: autoray couldn't find function 'get_dtype_name' for backend 'autograd'.

```

### System information

```shell
*
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes #4492 [sc-43886]

The autoray version of `get_dtype_name` breaks with array boxes. This patch makes it so that `get_dtype_name` is called on the value of the array box.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -357,6 +357,8 @@ array([False, False])
 
 <h3>Bug fixes 🐛</h3>
 
+* `qml.math.get_dtype_name` now works with autograd array boxes.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:360]:
```suggestion
* `qml.math.get_dtype_name` now works with autograd array boxes.
  [(#4494)](https://github.com/PennyLaneAI/pennylane/pull/4494)
```

**Review by mudit2812 — APPROVED** [review]

File: tests/math/test_functions.py [hunk 2]
```diff
@@ -1327,6 +1327,28 @@ def test_shape(shape, interface, create_array):
...
 
+@pytest.mark.parametrize(
+    "x, expected",
+    (
+        (1.0, "float64"),
+        (1, "int64"),
+        (onp.array(0.5), "float64"),
+        (onp.array(1.0, dtype="float32"), "float32"),
+        (ArrayBox(1, "a", "b"), "int64"),
+        (np.array(0.5, dtype="complex64"), "complex64")
```
Comment (timmysilv) [inline tests/math/test_functions.py:1338]:
```suggestion
        (np.array(0.5, dtype="complex64"), "complex64"),
```

File: tests/math/test_functions.py [hunk 3]
```diff
@@ -1327,6 +1327,28 @@ def test_shape(shape, interface, create_array):
...
 
 
+@pytest.mark.parametrize(
+    "x, expected",
+    (
+        (1.0, "float64"),
+        (1, "int64"),
+        (onp.array(0.5), "float64"),
+        (onp.array(1.0, dtype="float32"), "float32"),
+        (ArrayBox(1, "a", "b"), "int64"),
```
Comment (albi3ro) [inline tests/math/test_functions.py:1337]:
```suggestion
        (ArrayBox(1, "a", "b"), "int64"),
        (np.array(0.5), "float64"),
```

**Review by timmysilv — APPROVED** [review]
🎉

**Pull request merged**